cypress/integration/cucumber/step_definitions/refapp-3.x/01-login/login.js

```diff
@@ -0,0 +1,35 @@
+import {Given} from 'cypress-cucumber-preprocessor/steps';
+
+Given('user arrives at the login page', () => {
+    cy.visitPage('/login');
+})
+
+When('the user logs in with {string} and {string} to the {string}', (username, password, location) => {
+    cy.get('#username').type(username);
```

Comment:
Don't get elements by ID. IDs are implementation details. Get elements by how they appear to the user. In this case, it is a text input with label `Username`.

Reply:
do you have a plan to maintain the ids permanent? if so then i would strongly think ids/classes etc are a great ways to refer to elements vs labels which may change with locale changes etc which could simply be out of the test scope

Reply:
No; ids and classes are implementation details. "Implementation details" are aspects of the code which are invisible to the user. [Do not test implementation details](https://kentcdodds.com/blog/testing-implementation-details#why-is-testing-implementation-details-bad). We should test against the English labels. Those are the "source" labels in the code. If those change, it will be highly apparent, and easy to adapt the tests. AFAIK locale changes are not currently in test scope; but testing them is not out of the question.

Reply:
This is why I have recommended we use [Cypress Testing Library](https://testing-library.com/docs/cypress-testing-library/intro/)—it provides very natural semantics for "testing from the user's perspective." See also the [Testing Library Introduction](https://testing-library.com/docs/). This video, "[Build invincible tests](https://www.youtube.com/watch?v=F8LH9d9eN3M)," is also very good.